File: web/src/features/usage-logs/lib/model-token-adjustment.ts
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
import type { UsageLog } from '../data/schema'
import type { LogOtherData, ModelTokenUsage } from '../types'

export type TokenUsageDisplay = {
  displayed: ModelTokenUsage
  actual: ModelTokenUsage
  hasAdjustment: boolean
}

function cacheCreationTotal(usage: ModelTokenUsage) {
  const splitTotal =
    (usage.cache_creation_5m || 0) + (usage.cache_creation_1h || 0)
  return splitTotal > 0 ? splitTotal : usage.cache_creation || 0
}

function normalizeTokenUsage(usage: ModelTokenUsage): ModelTokenUsage {
  return {
    input: usage.input || 0,
    output: usage.output || 0,
    cache_read: usage.cache_read || 0,
    cache_creation: cacheCreationTotal(usage),
    cache_creation_5m: usage.cache_creation_5m || 0,
    cache_creation_1h: usage.cache_creation_1h || 0,
  }
}

export function getTokenUsageDisplay(
  log: UsageLog,
  other: LogOtherData | null,
  isAdmin: boolean
): TokenUsageDisplay {
  const fallback = normalizeTokenUsage({
    input: log.prompt_tokens || 0,
    output: log.completion_tokens || 0,
    cache_read: other?.cache_tokens || 0,
    cache_creation: other?.cache_creation_tokens || 0,
    cache_creation_5m: other?.cache_creation_tokens_5m || 0,
    cache_creation_1h: other?.cache_creation_tokens_1h || 0,
  })
  const audit = isAdmin ? other?.admin_info?.model_token_adjustment : undefined
  if (!audit) {
    return { displayed: fallback, actual: fallback, hasAdjustment: false }
  }

  return {
    displayed: normalizeTokenUsage(audit.billed),
    actual: normalizeTokenUsage(audit.actual),
    hasAdjustment: true,
  }
}

File: types/model_token_adjustment.go
package types

import "math"

const MaxModelTokenAdjustment = 100

type ModelTokenAdjustment struct {
	Input         *float64 `json:"input,omitempty"`
	Output        *float64 `json:"output,omitempty"`
	CacheRead     *float64 `json:"cache_read,omitempty"`
	CacheCreation *float64 `json:"cache_creation,omitempty"`
}

func (a ModelTokenAdjustment) HasAny() bool {
	return a.Input != nil || a.Output != nil || a.CacheRead != nil || a.CacheCreation != nil
}

func (a ModelTokenAdjustment) InputMultiplier() float64 {
	return modelTokenMultiplier(a.Input)
}

func (a ModelTokenAdjustment) OutputMultiplier() float64 {
	return modelTokenMultiplier(a.Output)
}

func (a ModelTokenAdjustment) CacheReadMultiplier() float64 {
	return modelTokenMultiplier(a.CacheRead)
}

func (a ModelTokenAdjustment) CacheCreationMultiplier() float64 {
	return modelTokenMultiplier(a.CacheCreation)
}

func modelTokenMultiplier(adjustment *float64) float64 {
	if adjustment == nil || math.IsNaN(*adjustment) || math.IsInf(*adjustment, 0) || *adjustment < 0 || *adjustment > MaxModelTokenAdjustment {
		return 1
	}
	return 1 + *adjustment
}

type ModelTokenUsage struct {
	Input           int `json:"input"`
	Output          int `json:"output"`
	CacheRead       int `json:"cache_read,omitempty"`
	CacheCreation   int `json:"cache_creation,omitempty"`
	CacheCreation5m int `json:"cache_creation_5m,omitempty"`
	CacheCreation1h int `json:"cache_creation_1h,omitempty"`
}

type ModelTokenAdjustmentAudit struct {
	Adjustments ModelTokenAdjustment `json:"adjustments"`
	Actual      ModelTokenUsage      `json:"actual"`
	Billed      ModelTokenUsage      `json:"billed"`
}
